pkg/chunk/cassandra/scanner.go

```diff
@@ -127,20 +136,30 @@ func (i *IndexValidator) checkEntry(
 		return
 	}
 
-	var exists int
-	err = i.o.readSession.Query(
-		fmt.Sprintf("SELECT count(*) FROM %s WHERE hash = ?", chunkTable),
-		c.ExternalKey(),
-	).WithContext(ctx).Scan(&exists)
-
-	if err != nil {
-		fmt.Println(err)
-		return
+	var chunkExists, ok bool
+	i.chunkIDMapMutex.RLock()
+	chunkExists, ok = i.chunkIDMap[chunkID]
+	i.chunkIDMapMutex.RUnlock()
+
+	if !ok {
+		var count int
+		err = i.o.readSession.Query(
+			fmt.Sprintf("SELECT count(*) FROM %s WHERE hash = ?", chunkTable),
+			c.ExternalKey(),
+		).WithContext(ctx).Scan(&count)
+
+		if err != nil {
+			fmt.Println(err)
+			return
+		}
+		chunkExists = count > 0
+		i.chunkIDMapMutex.Lock()
+		i.chunkIDMap[chunkID] = chunkExists
+		i.chunkIDMapMutex.Unlock()
 	}
-	if exists == 0 {
-		logrus.WithField("chunk_id", chunkID).Infoln("chunk not found, adding to output file")
+
+	if !chunkExists {
```

Comment:
what's the purpose of `chunkIDMap`? Is it to prevent that the same chunkID gets checked multiple times, because multiple index entries might refer to the same chunkID?

If multiple index entries refer to the same chunk and this chunk is absent, then the first time one of the entries gets scanned `i.chunkIDMap` is going to get the key `chunkID` set with the value `false`. The second time the key `chunkID` gets looked up from the map on `:141`, then `chunkExists` gets set to `false`, which will then push that same `chunkID` into the `out` channel a second time, no? not sure if i'm misunderstanding what the `chunkID` represents.